File: src/leetcode/python/IntegerBreak_343.py
class IntegerBreak_343(object):
    def integerBreak(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [0] * (n+1)
        dp[1] = 1
        for i in range(2, n+1):
            point = i // 2
            dp[i] = max(max(dp[point], point) * max(dp[i-point], i-point),
                        max(dp[point-1], point-1) * max(dp[i-point+1], i-point+1))
        return dp[n]

File: src/leetcode/java/IntegerBreak_343.java
package leetcode.java;

public class IntegerBreak_343 {
  public int integerBreak(int n) {
    int[] dp = new int[n + 1];
    dp[1] = 1;
    int breakPoint;
    for (int i = 2; i <= n; i++) {
      breakPoint = i / 2;
      dp[i] =
          Math.max(
            Math.max(dp[breakPoint], breakPoint) * Math.max(dp[i - breakPoint], i - breakPoint),
            Math.max(dp[breakPoint - 1], breakPoint - 1) * Math.max(dp[i - breakPoint + 1], i - breakPoint + 1));
    }
    return dp[n];
  }
}
